plugins/aggregators/histogram/histogram.go

```diff
@@ -0,0 +1,244 @@
+	tags                map[string]string
+}
+
+// counts is the number of hits in the bucket
+type counts []int64
+
+// NewHistogramAggregator creates new histogram aggregator
+func NewHistogramAggregator() telegraf.Aggregator {
+	h := &HistogramAggregator{}
+	h.buckets = make(bucketsByMetrics)
+	h.resetCache()
+
+	return h
+}
+
+// sampleConfig is config sample of histogram aggregation plugin
+var sampleConfig = `
+  # # Configuration for aggregate histogram metrics
+  # [[aggregators.histogram]]
+  #   ## General Aggregator Arguments:
+  #   ## The period on which to flush & clear the aggregator.
+  #   period = "30s"
+  #   ## If true, the original metric will be dropped by the
+  #   ## aggregator and will not get sent to the output plugins.
+  #   drop_original = false
+  #
+  #   ## The example of config to aggregate histogram for all fields of specified metric.
+  #   [[aggregators.histogram.config]]
+  #     ## The set of buckets.
+  #     buckets = [0.0, 15.6, 34.5, 49.1, 71.5, 80.5, 94.5, 100.0]
+  #     ## The name of metric.
+  #     metric_name = "cpu"
+  #
+  #   ## The example of config to aggregate for specified fields of metric.
+  #   [[aggregators.histogram.config]]
+  #     ## The set of buckets.
+  #     buckets = [0.0, 10.0, 20.0, 30.0, 40.0, 50.0, 60.0, 70.0, 80.0, 90.0, 100.0]
+  #     ## The name of metric.
+  #     metric_name = "diskio"
+  #     ## The concrete fields of metric
+  #     metric_fields = ["io_time", "read_time", "write_time"]
+`
+
+// SampleConfig returns sample of config
+func (h *HistogramAggregator) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns description of aggregator plugin
+func (h *HistogramAggregator) Description() string {
+	return "Keep the aggregate histogram of each metric passing through."
+}
+
+// Add adds new hit to the buckets
+func (h *HistogramAggregator) Add(in telegraf.Metric) {
+	id := in.HashID()
+	agr, ok := h.cache[id]
+	if !ok {
+		agr = metricHistogramCollection{
+			metric:              in.Name(),
+			tags:                in.Tags(),
+			histogramCollection: make(map[string]counts),
+		}
+	}
+
+	for field, value := range in.Fields() {
+		buckets := h.getBuckets(in.Name(), field)
+		if buckets == nil {
+			continue
+		}
+
+		if agr.histogramCollection[field] == nil {
+			agr.histogramCollection[field] = make(counts, len(buckets)+1)
+		}
+
+		if value, ok := convert(value); ok {
+			index := sort.SearchFloat64s(buckets, value)
+			agr.histogramCollection[field][index]++
+		}
+	}
+
+	h.cache[id] = agr
+}
+
+// Push returns histogram values for metrics
+func (h *HistogramAggregator) Push(acc telegraf.Accumulator) {
+	for _, aggregate := range h.cache {
+		for field, counts := range aggregate.histogramCollection {
+
+			buckets := h.getBuckets(aggregate.metric, field)
+			count := int64(0)
+
+			for index, bucket := range buckets {
+				count += counts[index]
+				addFields(acc, aggregate, field, strconv.FormatFloat(bucket, 'f', 1, 64), count)
```

Comment:
I agree it would be confusing. What do you think about specifying the buckets as a list of strings, a little more annoying to type in but then we could report exactly as inputted.